File: libs/east-c/packages/east-c/tests/test_beast2_manifest.c
/*
 * Segment manifests, pinned against TypeScript's manifest writer: the
 * directories three 50,000-element collections (and an empty one) are written
 * as are TypeScript's to the byte — the manifest's SHA-256 is pinned, and the
 * manifest names every segment by its own — each object is named by its
 * SHA-256, and a directory reads back, lazily through the manifest pager or
 * whole, as the value it was written from. Run under ASan/LSan for the pager's
 * segment source and the writer's lifetimes, their error paths included.
 */

#include <east/compat.h>
#include <east/east.h>
#include <east/type_of_type.h>

#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static int failures = 0;

#define CHECK(cond, ...)                                                                           \
    do {                                                                                           \
        if (!(cond)) {                                                                             \
            fprintf(stderr, "FAIL %s:%d: ", __FILE__, __LINE__);                                   \
            fprintf(stderr, __VA_ARGS__);                                                          \
            fprintf(stderr, "\n");                                                                 \
            failures++;                                                                            \
        }                                                                                          \
    } while (0)

/* TypeScript's manifests of the parity values below — the SHA-256 of each
 * manifest's bytes, as Beast2ManifestWriter writes them
 * (`east/src/serialization/beast2/v5/manifest-writer.spec.ts`). */
#define DICT_MANIFEST "5564111725b1962b7ae8c82cf91d24c4e1fb8d2c9613245693757fe4ff4187b0"
#define SET_MANIFEST "7b2c57db1da82f98ef88018589dc7a4f96bac120472b35e9826dbd60c395789c"
#define ARRAY_MANIFEST "127030afceae0c9590ad26ed26521cff170b8024229f28ccd54d647fc218df4c"
#define EMPTY_DICT_MANIFEST "356518ed344e3e210d0a2acb26eef7eff25cd76ba08b1aedd0640af5256cefb0"

static char g_dir[64];

/* ----- files ------------------------------------------------------------ */

static ByteBuffer *read_file(const char *path)
{
    FILE *f = fopen(path, "rb");
    if (!f) return NULL;
    fseek(f, 0, SEEK_END);
    long len = ftell(f);
    fseek(f, 0, SEEK_SET);
    ByteBuffer *buf = byte_buffer_new(len > 0 ? (size_t)len : 1);
    buf->len = fread(buf->data, 1, (size_t)len, f);
    fclose(f);
    return buf;
}

static void path_of(char *out, size_t cap, const char *name)
{
    snprintf(out, cap, "%s/%s", g_dir, name);
}

static void object_path(char *out, size_t cap, const char *manifest_path, const char *hash,
                        size_t len)
{
    snprintf(out, cap, "%s.segments/%.*s.beast2", manifest_path, (int)len, hash);
}

static EastValue *entries_of(EastValue *manifest)
{
    return east_struct_get_field_idx(manifest, 5);
}

static EastValue *entry_hash(EastValue *manifest, size_t i)
{
    return east_struct_get_field_idx(east_array_get(entries_of(manifest), i), 0);
}

/* The manifest a directory's file holds, decoded. */
static EastValue *manifest_at(const char *path)
{
    ByteBuffer *bytes = read_file(path);
    EastValue *manifest = NULL;
    int found = bytes ? east_beast2_read_manifest(bytes->data, bytes->len, &manifest) : 0;
    CHECK(found == 1, "%s does not hold a manifest", path);
    byte_buffer_free(bytes);
    return manifest;
}

/* Removes a manifest directory: every object the manifest names, the objects
 * directory, and the manifest. */
static void remove_dir(const char *path)
{
    EastValue *manifest = manifest_at(path);
    char object[512];
    if (manifest) {
        EastValue *header = east_struct_get_field_idx(manifest, 4);
        object_path(object, sizeof(object), path, header->data.string.data,
                    header->data.string.len);
        remove(object);
        for (size_t i = 0; i < east_array_len(entries_of(manifest)); i++) {
            EastValue *hash = entry_hash(manifest, i);
            object_path(object, sizeof(object), path, hash->data.string.data,
                        hash->data.string.len);
            remove(object);
        }
        east_value_release(manifest);
    }
    snprintf(object, sizeof(object), "%s.segments", path);
    rmdir(object);
    remove(path);
}

/* The manifest file's SHA-256, and every object's name its own SHA-256. */
static void check_directory(const char *what, const char *path, const char *pin,
                            size_t expected_entries)
{
    ByteBuffer *bytes = read_file(path);
    CHECK(bytes != NULL, "%s: no manifest written", what);
    if (!bytes) return;
    char hex[EAST_SHA256_HEX_SIZE];
    east_sha256_hex(bytes->data, bytes->len, hex);
    CHECK(strcmp(hex, pin) == 0, "%s: the manifest diverges from TypeScript's: %s (expected %s)",
          what, hex, pin);
    byte_buffer_free(bytes);

    EastValue *manifest = manifest_at(path);
    if (!manifest) return;
    size_t n = east_array_len(entries_of(manifest));
    CHECK(n == expected_entries, "%s: %zu entries, expected %zu", what, n, expected_entries);
    char object[512];
    for (size_t i = 0; i <= n; i++) {
        /* Entry i's segment, and after the entries the header. */
        EastValue *hash = i < n ? entry_hash(manifest, i) : east_struct_get_field_idx(manifest, 4);
        object_path(object, sizeof(object), path, hash->data.string.data, hash->data.string.len);
        ByteBuffer *blob = read_file(object);
        CHECK(blob != NULL, "%s: object %.64s is missing", what, hash->data.string.data);
        if (!blob) continue;
        east_sha256_hex(blob->data, blob->len, hex);
        CHECK(strncmp(hex, hash->data.string.data, 64) == 0, "%s: object %.64s hashes to %s", what,
              hash->data.string.data, hex);
        byte_buffer_free(blob);
    }
    east_value_release(manifest);
}

/* ----- values ------------------------------------------------------------ */

static EastValue *parity_dict(EastType *type, size_t n)
{
    EastValue *d = east_dict_new(type->data.dict.key, type->data.dict.value);
    for (size_t i = 0; i < n; i++) {
        char buf[16];
        snprintf(buf, sizeof(buf), "k%07u", (unsigned)i);
        EastValue *k = east_string(buf), *v = east_integer((int64_t)i);
        east_dict_set(d, k, v);
        east_value_release(k);
        east_value_release(v);
    }
    return d;
}

static EastValue *parity_set(EastType *type)
{
    EastValue *s = east_set_new(type->data.element);
    for (size_t i = 0; i < 50000; i++) {
        char buf[16];
        snprintf(buf, sizeof(buf), "e%07u", (unsigned)i);
        EastValue *e = east_string(buf);
        east_set_insert(s, e);
        east_value_release(e);
    }
    return s;
}

static EastValue *parity_array(EastType *type)
{
    EastValue *a = east_array_new(type->data.element);
    for (size_t i = 0; i < 50000; i++) {
        char buf[16];
        snprintf(buf, sizeof(buf), "a%07u", (unsigned)i);
        EastValue *e = east_string(buf);
        east_array_push(a, e);
        east_value_release(e);
    }
    return a;
}

/* ----- the gates ---------------------------------------------------------- */

static void test_parity_directories(void)
{
    char path[256];
    EastType *dict_type = east_dict_type(&east_string_type, &east_integer_type);
    EastValue *dict = parity_dict(dict_type, 50000);
    path_of(path, sizeof(path), "dict.beast2");
    CHECK(east_beast2_write_manifest_dir(dict, dict_type, EAST_BEAST2_CODEC_DEFLATE, path),
          "writing the Dict failed: %s", east_builtin_get_error());
    check_directory("Dict", path, DICT_MANIFEST, 38);
    remove_dir(path);
    east_value_release(dict);

    EastType *set_type = east_set_type(&east_string_type);
    EastValue *set = parity_set(set_type);
    path_of(path, sizeof(path), "set.beast2");
    CHECK(east_beast2_write_manifest_dir(set, set_type, EAST_BEAST2_CODEC_DEFLATE, path),
          "writing the Set failed: %s", east_builtin_get_error());
    check_directory("Set", path, SET_MANIFEST, 44);
    remove_dir(path);
    east_value_release(set);

    EastType *array_type = east_array_type(&east_string_type);
    EastValue *array = parity_array(array_type);
    path_of(path, sizeof(path), "array.beast2");
    CHECK(east_beast2_write_manifest_dir(array, array_type, EAST_BEAST2_CODEC_DEFLATE, path),
          "writing the Array failed: %s", east_builtin_get_error());
    check_directory("Array", path, ARRAY_MANIFEST, 43);
    remove_dir(path);
    east_value_release(array);

    EastValue *empty = east_dict_new(&east_string_type, &east_integer_type);
    path_of(path, sizeof(path), "empty.beast2");
    CHECK(east_beast2_write_manifest_dir(empty, dict_type, EAST_BEAST2_CODEC_DEFLATE, path),
          "writing the empty Dict failed: %s", east_builtin_get_error());
    check_directory("empty Dict", path, EMPTY_DICT_MANIFEST, 0);
    EastValue *manifest = manifest_at(path);
    EastValue *back =
        manifest ? east_beast2_decode_manifest_dir(path, manifest, dict_type, true) : NULL;
    CHECK(back && east_dict_len(back) == 0, "the empty Dict does not read back empty");
    if (back) east_value_release(back);
    if (manifest) east_value_release(manifest);
    remove_dir(path);
    east_value_release(empty);
}

/* A directory reads back as its value: lazily, a keyed read opening the one
 * segment it lands in; and whole, through a hydrate or a decode. */
static void test_read_back(void)
{
    char path[256];
    EastType *type = east_dict_type(&east_string_type, &east_integer_type);
    EastValue *dict = parity_dict(type, 50000);
    path_of(path, sizeof(path), "read.beast2");
    CHECK(east_beast2_write_manifest_dir(dict, type, EAST_BEAST2_CODEC_DEFLATE, path),
          "writing failed: %s", east_builtin_get_error());
    EastValue *manifest = manifest_at(path);
    if (!manifest) {
        east_value_release(dict);
        return;
    }

    EastValue *lazy = east_beast2_open_manifest_dir(path, manifest, type, true);
    CHECK(lazy && lazy->kind == EAST_VAL_PAGED, "the directory did not open lazily: %s",
          east_builtin_get_error());
    if (lazy) {
        CHECK(east_dict_len(lazy) == 50000, "the lazy Dict holds %zu pairs", east_dict_len(lazy));
        Beast2Pages *pages = lazy->data.paged.pages;
        EastValue *key = east_string("k0031415");
        EastValue *value = NULL;
        CHECK(east_beast2_pages_get_key(pages, key, &value) == 1 && value &&
                  value->data.integer == 31415,
              "a keyed read of the directory missed");
        if (value) east_value_release(value);
        east_value_release(key);
        size_t decoded = 0, probed = 0;
        east_beast2_pages_stats(pages, &decoded, &probed);
        CHECK(decoded == 1, "a keyed read decoded %zu segments, not one", decoded);

        EastValue *whole = east_paged_hydrated(lazy);
        CHECK(whole && east_value_equal(whole, dict), "the hydrated directory is not the value");
        east_value_release(lazy);
    }

    EastValue *decoded = east_beast2_decode_manifest_dir(path, manifest, type, false);
    CHECK(decoded && east_value_equal(decoded, dict), "the decoded directory is not the value");
    if (decoded) east_value_release(decoded);

    /* A segment gone from the directory is named when a read needs it. */
    char object[512];
    EastValue *hash = entry_hash(manifest, 7);
    object_path(object, sizeof(object), path, hash->data.string.data, hash->data.string.len);
    remove(object);
    CHECK(east_beast2_decode_manifest_dir(path, manifest, type, false) == NULL,
          "a directory missing a segment decoded");
    char *err = east_builtin_get_error();
    CHECK(err && strstr(err, "cannot be read") && strstr(err, hash->data.string.data),
          "the missing segment was reported as: %s", err ? err : "nothing");
    free(err);

    remove_dir(path);
    east_value_release(manifest);
    east_value_release(dict);
}

/* What is not a manifest is not read as one. */
static void test_recognition(void)
{
    EastType *type = east_dict_type(&east_string_type, &east_integer_type);
    EastValue *dict = parity_dict(type, 10);
    ByteBuffer *blob = east_beast2_encode_paged(dict, type, EAST_BEAST2_CODEC_DEFLATE);
    EastValue *manifest = NULL;
    CHECK(east_beast2_read_manifest(blob->data, blob->len, &manifest) == 0 && !manifest,
          "a collection blob read as a manifest");
    static const uint8_t junk[16] = {1, 2, 3};
    CHECK(east_beast2_read_manifest(junk, sizeof(junk), &manifest) == 0,
          "sixteen bytes of junk read as a manifest");
    byte_buffer_free(blob);
    east_value_release(dict);
}

/* ----- a sink that refuses ------------------------------------------------ */

typedef struct {
    size_t objects;
    size_t refuse_at; /* the object refused, counting the header as 0 */
    bool manifest_written;
} Refusing;

static bool refusing_object(void *ctx, const char *hash, const uint8_t *bytes, size_t len)
{
    (void)hash;
    (void)bytes;
    (void)len;
    Refusing *r = ctx;
    if (r->objects++ == r->refuse_at) {
        east_builtin_error("test: the store is full");
        return false;
    }
    return true;
}

static bool refusing_manifest(void *ctx, const uint8_t *bytes, size_t len)
{
    (void)bytes;
    (void)len;
    ((Refusing *)ctx)->manifest_written = true;
    return true;
}

/* A segment the sink refuses fails the write, and no manifest is written. */
static void test_refused_segment(void)
{
    EastType *type = east_dict_type(&east_string_type, &east_integer_type);
    EastValue *dict = parity_dict(type, 50000);
    Refusing r = {0, 3, false};
    Beast2ManifestSink sink = {&r, refusing_object, refusing_manifest};
    Beast2ManifestWriter *w =
        east_beast2_manifest_writer_new(type, EAST_BEAST2_CODEC_DEFLATE, &sink);
    bool ok = w != NULL;
    for (size_t i = 0; ok && i < east_dict_len(dict); i++)
        ok = east_beast2_manifest_writer_add_pair(w, east_dict_key_at(dict, i),
                                                  east_dict_val_at(dict, i));
    char *err = east_builtin_get_error();
    CHECK(!ok && err && strcmp(err, "test: the store is full") == 0, "the refusal surfaced as: %s",
          err ? err : "nothing");
    free(err);
    CHECK(w && !east_beast2_manifest_writer_add_pair(w, east_dict_key_at(dict, 0),
                                                     east_dict_val_at(dict, 0)),
          "the writer took an element after a refused segment");
    free(east_builtin_get_error());
    CHECK(w && !east_beast2_manifest_writer_finish(w) && !r.manifest_written,
          "a failed writer wrote its manifest");
    free(east_builtin_get_error());
    east_beast2_manifest_writer_free(w);
    east_value_release(dict);
}

int main(void)
{
    east_type_of_type_init();
    BuiltinRegistry *builtins = builtin_registry_new();
    east_register_all_builtins(builtins);
    PlatformRegistry *platform = platform_registry_new();
    east_set_thread_context(platform, builtins);

    snprintf(g_dir, sizeof(g_dir), "manifest_gate_XXXXXX");
    if (!mkdtemp(g_dir)) {
        fprintf(stderr, "cannot create a scratch directory\n");
        return 1;
    }
    test_parity_directories();
    test_read_back();
    test_recognition();
    test_refused_segment();
    rmdir(g_dir);

    platform_registry_free(platform);
    builtin_registry_free(builtins);
    if (failures > 0) {
        fprintf(stderr, "%d check(s) failed\n", failures);
        return 1;
    }
    printf("beast2 manifests: all checks passed\n");
    return 0;
}

File: libs/east-py/packages/east-py/tests/serialization/test_beast2_manifest.py
#
# Copyright (c) 2025 Elara AI Pty Ltd
# Licensed under the Business Source License 1.1. See LICENSE.md for details.
#
"""Segment manifests, pinned against TypeScript's manifest writer.

A manifest directory — the manifest's file and, in ``<file>.segments/``, the
header and every segment it names as ``<sha256>.beast2`` — is the form e3
stores a collection in, and the one a runner reads a staged input from. This
runtime binds east-c's writer and reader, so what is asserted here is that the
directories python writes are TypeScript's to the byte
(``east/src/serialization/beast2/v5/manifest-writer.spec.ts``) and east-c's
(``tests/test_beast2_manifest.c``), that they read back as their values —
whole, and lazily through the manifest pager — and that the merge reads and
writes them. The fixtures and the pins are the same on all three sides.
"""

import hashlib
from pathlib import Path

import pytest
from east.runtime._compiler_eastc import open_manifest_file, paged_value_stats
from east.serialization._beast2_eastc import _encode_beast2_fence, _merge_blobs

from east import ArrayType, DictType, East, EastDict, EastSet, IntegerType, SetType, StringType
from east.serialization.beast2 import (
    SEGMENT_RULE_ARRAY,
    SEGMENT_RULE_KEYED,
    Beast2ManifestWriter,
    encode_beast2_paged_for,
    load_beast2_manifest,
    read_beast2_manifest,
)

#: TypeScript's manifests of the parity values below — the SHA-256 of each
#: manifest's bytes, as its Beast2ManifestWriter writes them.
DICT_MANIFEST = "5564111725b1962b7ae8c82cf91d24c4e1fb8d2c9613245693757fe4ff4187b0"
SET_MANIFEST = "7b2c57db1da82f98ef88018589dc7a4f96bac120472b35e9826dbd60c395789c"
ARRAY_MANIFEST = "127030afceae0c9590ad26ed26521cff170b8024229f28ccd54d647fc218df4c"
EMPTY_DICT_MANIFEST = "356518ed344e3e210d0a2acb26eef7eff25cd76ba08b1aedd0640af5256cefb0"

TABLE = DictType(StringType, IntegerType)


def _parity_dict() -> EastDict:
    d = EastDict(StringType, IntegerType)
    for i in range(50_000):
        d[f"k{i:07d}"] = i
    return d


@pytest.mark.parametrize(
    ("collection_type", "value", "pin", "entries", "rule"),
    [
        (TABLE, _parity_dict, DICT_MANIFEST, 38, SEGMENT_RULE_KEYED),
        (SetType(StringType), lambda: EastSet(StringType, [f"e{i:07d}" for i in range(50_000)]),
         SET_MANIFEST, 44, SEGMENT_RULE_KEYED),
        (ArrayType(StringType), lambda: [f"a{i:07d}" for i in range(50_000)],
         ARRAY_MANIFEST, 43, SEGMENT_RULE_ARRAY),
        (TABLE, lambda: EastDict(StringType, IntegerType), EMPTY_DICT_MANIFEST, 0, SEGMENT_RULE_KEYED),
    ],
    ids=["Dict", "Set", "Array", "empty Dict"],
)
def test_a_directory_is_the_one_typescript_writes(tmp_path, collection_type, value, pin, entries, rule):
    path = tmp_path / "value.beast2"
    with Beast2ManifestWriter(collection_type, path) as writer:
        writer.add_all(value())
    assert hashlib.sha256(path.read_bytes()).hexdigest() == pin

    manifest = read_beast2_manifest(path)
    assert (manifest["kind"], manifest["level"], manifest["rule"]) == ("$segments", 0, rule)
    assert len(manifest["entries"]) == entries
    # Every object — each segment, and the header they are written under — is
    # named by its own SHA-256, and the directory holds nothing else.
    names = [entry["hash"] for entry in manifest["entries"]] + [manifest["header"]]
    objects = Path(f"{path}.segments")
    for name in names:
        assert hashlib.sha256((objects / f"{name}.beast2").read_bytes()).hexdigest() == name
    assert sorted(p.name for p in objects.iterdir()) == sorted({f"{name}.beast2" for name in names})


def test_the_entries_carry_each_segments_count_size_and_first_key(tmp_path):
    path = tmp_path / "dict.beast2"
    with Beast2ManifestWriter(TABLE, path) as writer:
        writer.add_all(_parity_dict())
    manifest = read_beast2_manifest(path)
    start = 0
    for entry in manifest["entries"]:
        segment = Path(f"{path}.segments") / f"{entry['hash']}.beast2"
        assert entry["bytes"] == segment.stat().st_size
        assert entry["fence"] == _encode_beast2_fence(StringType, f"k{start:07d}")
        start += entry["count"]
    assert start == 50_000


def test_the_writer_writes_the_same_directory_however_it_is_fed(tmp_path):
    value = _parity_dict()
    path = tmp_path / "dict.beast2"
    with Beast2ManifestWriter(TABLE, path) as writer:
        for entry in list(value.items())[:10_000]:
            writer.add(entry)
        rest = EastDict(StringType, IntegerType)
        for key, number in list(value.items())[10_000:]:
            rest[key] = number
        writer.add_all(rest)
    assert hashlib.sha256(path.read_bytes()).hexdigest() == DICT_MANIFEST


def test_a_directory_reads_back_as_its_value(tmp_path):
    value = _parity_dict()
    path = tmp_path / "dict.beast2"
    with Beast2ManifestWriter(TABLE, path) as writer:
        writer.add_all(value)
    # A manifest records its type, so none need be given; one given must be it.
    assert dict(load_beast2_manifest(path).items()) == dict(value.items())
    assert dict(load_beast2_manifest(path, TABLE).items()) == dict(value.items())
    with pytest.raises(ValueError, match="declared type does not match the manifest"):
        load_beast2_manifest(path, DictType(StringType, StringType))


def test_a_directory_opens_lazily_and_a_keyed_read_decodes_one_segment(tmp_path):
    path = tmp_path / "dict.beast2"
    with Beast2ManifestWriter(TABLE, path) as writer:
        writer.add_all(_parity_dict())
    hold = open_manifest_file(TABLE, path)
    assert hold is not None
    read = East.compile(East.function([TABLE], IntegerType,
                                      lambda _b, d: d.size() + d.get("k0031415")))
    assert read(hold) == 50_000 + 31_415
    segments, decoded, _fences, hydrated = paged_value_stats(hold._east_c_paged)
    assert (segments, decoded, hydrated) == (38, 1, False)

    with pytest.raises(ValueError, match="cannot open a blob of type"):
        open_manifest_file(DictType(StringType, StringType), path)


def test_a_missing_segment_is_named_when_a_read_needs_it(tmp_path):
    path = tmp_path / "dict.beast2"
    with Beast2ManifestWriter(TABLE, path) as writer:
        writer.add_all(_parity_dict())
    gone = read_beast2_manifest(path)["entries"][7]["hash"]
    (Path(f"{path}.segments") / f"{gone}.beast2").unlink()
    with pytest.raises(ValueError, match=rf"{gone}\.beast2 cannot be read"):
        load_beast2_manifest(path)


def test_what_is_not_a_manifest_is_not_read_as_one(tmp_path):
    blob = tmp_path / "blob.beast2"
    blob.write_bytes(encode_beast2_paged_for(TABLE)(EastDict(StringType, IntegerType, {"a": 1})))
    assert read_beast2_manifest(blob) is None
    assert read_beast2_manifest(blob.read_bytes()) is None
    assert read_beast2_manifest(bytes(16)) is None
    empty = tmp_path / "empty.beast2"
    empty.write_bytes(b"")
    assert read_beast2_manifest(empty) is None
    with pytest.raises(ValueError, match="does not hold a manifest"):
        load_beast2_manifest(blob)


def test_a_writer_left_by_an_exception_writes_no_manifest(tmp_path):
    path = tmp_path / "dict.beast2"
    with pytest.raises(RuntimeError, match="strictly ascending"), \
            Beast2ManifestWriter(TABLE, path) as writer:
        writer.add(("b", 1))
        writer.add(("a", 2))
    assert not path.exists()


def test_the_merge_reads_manifest_directories_and_writes_one(tmp_path):
    # The halves of the parity Dict, each a directory, merge into
    # TypeScript's directory of the whole — and into the whole's blob.
    value = _parity_dict()
    evens = EastDict(StringType, IntegerType)
    odds = EastDict(StringType, IntegerType)
    for i, (key, number) in enumerate(value.items()):
        (odds if i % 2 else evens)[key] = number
    halves = [tmp_path / "evens.beast2", tmp_path / "odds.beast2"]
    for half, part in zip(halves, (evens, odds), strict=True):
        with Beast2ManifestWriter(TABLE, half) as writer:
            writer.add_all(part)

    merged = tmp_path / "merged.beast2"
    assert _merge_blobs(halves, merged, output_manifest=True) == {
        "inputs": 2, "entries": 50_000, "folds": 0,
    }
    assert hashlib.sha256(merged.read_bytes()).hexdigest() == DICT_MANIFEST

    blob = tmp_path / "merged-blob.beast2"
    _merge_blobs(halves, blob)
    assert blob.read_bytes() == encode_beast2_paged_for(TABLE)(value)
